src/core/patterns.js

```diff
@@ -59,3 +59,61 @@ foam.CLASS({
     }
   ]
 });
+
+/** Causes an class to pool its instances. create() will pull from the pool,
+ and destroy() will return instances to the pool. Object pools can be found
+ in <code>foam.__objectPools__</code>. */
+foam.CLASS({
+  package: 'foam.pattern',
+  name: 'Pooled',
+  axioms: [ foam.pattern.Singleton.create() ],
+
+  methods: [
+    function installInClass(cls) {
+      // Keeping the object pools in an accessible location allows them
+      // to be cleared out.
+      if ( ! foam.__objectPools__ ) { foam.__objectPools__ = {}; }
+      var pool = foam.__objectPools__[cls];
+      if ( ! pool ) {
+        pool = foam.__objectPools__[cls] = [];
+      }
+
+      var oldCreate = cls.create;
+      cls.create = function(args, X) {
+        var nu;
+        // Pull from the pool, run the usual init process that .create() would
+        // do. TODO: Alter create to accept the base object so we don't duplicate
+        // this init code?
+        if ( pool.length ) {
+          nu = pool.splice(-1, 1)[0];
+          nu.destroyed = false;
+          nu.initArgs(args, X);
+          nu.init && nu.init();
+        } else {
+          nu = oldCreate.apply(this, arguments);
+        }
+        return nu;
+      }
+
+      var oldDestroy = cls.prototype.destroy;
+      cls.prototype.destroy = function() {
+        if ( this.destroyed ) return;
+
+        // Run destroy process on the object, but leave its privates empty but intact
+        // to avoid reallocating them
+        var inst_ = this.instance_;
```

Comment:
I feel like this cleanup shouldn't be here.

If an object needs to clean up in .destroy() it should already do so correctly.  The object pool should just add pooling of instances.  Cleanup should be a responsibility of the object.